### DISCUSSION TITLE
Frozen on M2 MacBook

### DISCUSSION THREAD

**Discussion by steve-abel** [MilesCranmer/PySR#391] [Q&A]
Hi there, 
     trying the basic test implementation at https://github.com/MilesCranmer/PySR/blob/master/examples/pysr_demo.ipynb
and when I get to model.fit(X, y) is hangs and says 

Compiling Julia backend...

for ever. What does this mean ? Is it doing anything? It has been like that for 10 mins. How long should it take normally to run on this test example? (some estimates would be useful). As I am on a jupyter notebook I understand that it is supposed to be printing out expressions every 5 seconds after the 1st iteration - so it hasn't yet done one iteration I think. If it makes a difference I am on an m2 mac.

Thanks for any help

Steve A.

**Comment by MilesCranmer** [comment]
Hi @steve-abel,

Since you are running in Jupyter, have you checked the Jupyter logs? It prints the logs to stderr so it might be printing to the log file depending on how you have Jupyter set up.

But in general the "Compiling Julia backend" should be quite short. On an M2 mac it should be like ~10 seconds. My guess is just that it's printing the rest of the search results somewhere other than the Jupyter output.

Cheers,
Miles

**Reply by steve-abel** [reply]
Thanks for your reply. Doesn't seem to be writing into the log file. I wonder if I have the right Julia installed. I tried the ARM 1.9 version and this complained about the architecture being wrong as soon as I load Julia. Then I tried the rosetta 1.9 version. This loads Julia OK but then stalls at the Julia backend step as before. I now also get (after Julia Beckend ...) the warning 
/Users/sabel/.venv_Genetic_Programming/lib/python3.9/site-packages/pysr/julia_helpers.py:231: UserWarning: Julia has already started. The new Julia options {'threads': 12} will be ignored.
  warnings.warn(

Any ideas. ? 

Steve

**Reply by steve-abel** [reply]
Ah - it seems to be writing a lot of Hall of fame pkl files. How do I view those? Using pickle.read  I get EOL while scanning. Is there a way to make it just output to the jupyter notebook? Sorry for all the dumb questions - I'm lost.

**Reply by MilesCranmer** [reply]
Can you run `example.py` from the PySR repository from the terminal? i.e., `python example.py`

**Reply by MilesCranmer** [reply]
To load the pickle files you can use `PySRRegressor.from_file`. See https://astroautomata.com/PySR/api/#pysr.sr.PySRRegressor.from_file

**Reply by steve-abel** [reply]
Hi there - so example.py takes about 4 minutes to go from Julia.backend ... to Started !
And then gets stuck. In jupyter it works one time in 2, and if I comment out redirect_output_streams() it produces cvs files and it finished if I take niterations=10 (but this seems a bit low). iterations = 20 seems to get stuck again. 
Does this run better on linux? Maybe I could install a virtual machine and do it there.

**Reply by steve-abel** [reply]
Is it possible the Julia installation using rosetta is screwing up?

**Reply by steve-abel** [reply]
python example.py was still going after 30 mins so I killed it

**Reply by MilesCranmer** [reply]
Oh, well Julia under Rosetta is going to be extremely slow unfortunately. Especially considering Rosetta will be applied to the various system libraries. Definitely use the ARM Julia.

macOS is perfectly fine. My personal hardware (where I do most PySR stuff) is an M1 MacBook, and it is pretty fast when running searches. I’ve gotten things working fine  in both brew and pyenv versions of pythons, so any of those should work. It’s only the Rosetta versions that are quite slow.

(Edit: actually just to mention for the Conda version, I use the global Julia and pip PySR, as there’s no ARM Julia available through conda)

**Reply by MilesCranmer** [reply]
Note that a good fallback option is docker if you just want something working quickly. It gives you native speeds so is actually fine for general use. See the instructions on the README for setting it up.

**Reply by steve-abel** [reply]
OK thanks - I think the problem is in my Julia installation then. I'll try to work that out. The programme did work but indeed very slowly.

**Reply by steve-abel** [reply]
Update - reinstalled my environments and it works really well now thanks --  Impressive I have to say!!

**Reply by MilesCranmer** [reply]
Awesome, happy to hear it!!
